src/gf.c

```diff
@@ -134,6 +142,13 @@ static inline int cache_match(jl_value_t **args, size_t n, jl_tupletype_t *sig,
                 // to match exactly either. this is cached as Type{T}.
                 // analogous to the situation with tuples.
             }
+            else if (jl_is_typector(a)) {
+                // in the case where a is exactly a
+                // TypeConstructor, tp0 must be exactly
+                // Type{TypeConstructor}
+                if (tp0 != (jl_value_t*)jl_typector_type)
```

Comment:
ok, these should not be here then. I wrote this first, but it only partially addressed the problem, so this might be unused now (after I added the code below).

might be good to manually verify that all 4 permutations of the following are working:
```
g11840(::DataType) = 1
g11840(::Type) = 2
@test g11840(Vector.body) == 1
@test g11840(Vector) == 2
```